### ISSUE TITLE
Implement MiniBatch K-Means for memory-efficient clustering

### ISSUE THREAD

**Issue by jxnl** [567-labs/kura#96]
## Problem Statement

The current K-Means implementation in Kura loads the entire embedding matrix into memory, which becomes problematic for large datasets. For 100k conversations with 1536-dimensional embeddings, this creates a matrix requiring ~1.2GB of memory just for the embeddings, not counting the overhead of the clustering algorithm itself.

## Current Implementation

The current implementation in `kura/k_means.py` uses scikit-learn's standard `KMeans`:

```python
# Line 49: Full matrix created
X = np.array(embeddings)
# Line 52: Entire dataset processed at once
kmeans = KMeans(n_clusters=n_clusters)
cluster_labels = kmeans.fit_predict(X)
```

### Limitations

1. **Memory Usage**: O(n × d) where n = number of items, d = embedding dimensions
2. **No Streaming**: Requires all data in memory before clustering begins
3. **No Progress Tracking**: Can't monitor progress for long-running operations
4. **Single-threaded**: Doesn't leverage mini-batch parallelization opportunities

### Benefits of MiniBatch K-Means

1. **Memory Efficiency**: Only loads small batches (e.g., 1024 samples) at a time
2. **Faster Convergence**: Often converges faster than standard K-Means on large datasets
3. **Online Learning**: Can process data streams or datasets larger than memory
4. **Progress Monitoring**: Natural checkpoints at each batch for progress tracking

#### 1. Create MiniBatchKMeansClusteringMethod

```python
from sklearn.cluster import MiniBatchKMeans
import numpy as np
from typing import TypeVar, Optional, Iterator
import logging

class MiniBatchKMeansClusteringMethod(BaseClusteringMethod):
    def __init__(
        self, 
        clusters_per_group: int = 10,
        batch_size: int = 1024,
        max_iter: int = 100,
        random_state: Optional[int] = 42,
        reassignment_ratio: float = 0.01,
        max_no_improvement: int = 10
    ):
        self.clusters_per_group = clusters_per_group
        self.batch_size = batch_size
        self.max_iter = max_iter
        self.random_state = random_state
        self.reassignment_ratio = reassignment_ratio
        self.max_no_improvement = max_no_improvement
        
    def cluster(self, items: list[T]) -> dict[int, list[T]]:
        n_clusters = math.ceil(len(items) / self.clusters_per_group)
        
        # Initialize MiniBatchKMeans
        mbkmeans = MiniBatchKMeans(
            n_clusters=n_clusters,
            batch_size=self.batch_size,
            max_iter=self.max_iter,
            random_state=self.random_state,
            reassignment_ratio=self.reassignment_ratio,
            max_no_improvement=self.max_no_improvement
        )
        
        # Process in batches
        embeddings_generator = self._batch_embeddings(items)
        for batch_embeddings in embeddings_generator:
            mbkmeans.partial_fit(batch_embeddings)
            
        # Final prediction on all data (in batches)
        cluster_labels = self._predict_in_batches(mbkmeans, items)
        
        return self._build_cluster_dict(items, cluster_labels)
```

#### 2. Add Streaming Support for Large Datasets

```python
def cluster_streaming(self, items_iterator: Iterator[T]) -> dict[int, list[T]]:
    """
    Cluster items from an iterator without loading all into memory.
    Useful for processing HuggingFace datasets in streaming mode.
    """
    # First pass: fit the model
    mbkmeans = self._fit_streaming(items_iterator)
    
    # Second pass: assign clusters
    # This requires a second pass but maintains memory efficiency
    return self._assign_clusters_streaming(mbkmeans, items_iterator)
```

#### 3. Add Progress Tracking

```python
from tqdm import tqdm

def cluster_with_progress(self, items: list[T]) -> dict[int, list[T]]:
    n_batches = math.ceil(len(items) / self.batch_size)
    
    with tqdm(total=n_batches, desc="Clustering") as pbar:
        for batch_idx, batch_embeddings in enumerate(self._batch_embeddings(items)):
            mbkmeans.partial_fit(batch_embeddings)
            
            # Log intermediate metrics
            if batch_idx % 10 == 0:
                inertia = mbkmeans.score(batch_embeddings)
                pbar.set_postfix({"inertia": f"{inertia:.2f}"})
            
            pbar.update(1)
```

#### 4. Memory-Mapped File Support

```python
def cluster_mmap(self, embeddings_file: str, items: list[T]) -> dict[int, list[T]]:
    """
    Use memory-mapped files for extremely large embedding matrices.
    Embeddings are stored on disk and accessed as needed.
    """
    # Load embeddings as memory-mapped array
    embeddings_mmap = np.memmap(
        embeddings_file, 
        dtype='float32', 
        mode='r',
        shape=(len(items), self.embedding_dim)
    )
    
    # Process in chunks without loading full matrix
    mbkmeans = MiniBatchKMeans(n_clusters=self.n_clusters)
    
    for i in range(0, len(items), self.batch_size):
        batch = embeddings_mmap[i:i + self.batch_size]
        mbkmeans.partial_fit(batch)
```

### Configuration Options

Add new configuration parameters to Kura:

```python
class KuraConfig:
    clustering_method: Literal["kmeans", "minibatch_kmeans"] = "minibatch_kmeans"
    clustering_batch_size: int = 1024  # For MiniBatch K-Means
    clustering_max_iter: int = 100
    clustering_reassignment_ratio: float = 0.01  # Fraction of centers to reassign
    use_mmap_for_large_datasets: bool = True  # Auto-enable for >50k items
```

### Performance Comparison

Expected improvements for 100k conversations:

| Metric | Standard K-Means | MiniBatch K-Means |
|--------|-----------------|-------------------|
| Memory Usage | ~1.2GB | ~8MB per batch |
| Time to First Cluster | After full fit | After first batch |
| Supports Streaming | No | Yes |
| Progress Tracking | No | Yes |
| Partial Fit | No | Yes |

### Migration Strategy

1. **Add as Alternative**: Keep existing K-Means, add MiniBatch as option
2. **Auto-Selection**: Use MiniBatch automatically for datasets > 10k items
3. **Gradual Default**: Make MiniBatch default in next major version

### Implementation Steps

1. Create `MiniBatchKMeansClusteringMethod` class
2. Add streaming support methods
3. Implement progress tracking
4. Add memory-mapped file support
5. Update configuration to support method selection
6. Add benchmarks comparing both methods
7. Update documentation with usage examples

### Testing Strategy

1. **Correctness Tests**: Ensure cluster quality is maintained
2. **Memory Tests**: Verify memory usage stays constant with dataset size
3. **Performance Tests**: Benchmark on various dataset sizes
4. **Streaming Tests**: Test with iterators and generators
5. **Edge Cases**: Empty batches, single item batches, etc.

### Success Criteria

- [ ] Memory usage remains constant regardless of dataset size
- [ ] Clustering quality (measured by silhouette score) within 5% of standard K-Means
- [ ] Support for streaming datasets from HuggingFace
- [ ] Progress tracking for long-running operations
- [ ] Benchmarks show >10x memory reduction for 100k+ datasets

### Related Components

- `kura/k_means.py`: Current K-Means implementation
- `kura/base_classes/clustering_method.py`: Base class to extend
- `kura/cluster.py`: Integration point for clustering methods
- `kura/types/cluster.py`: Cluster data structures

This enhancement is critical for supporting large-scale conversation analysis and aligns with the goal of processing 100k+ conversations efficiently.
